importer/src/path_canonicalizer.rs

```diff
@@ -0,0 +1,378 @@
+        )
+        .map(|(mut state, absolute_path, symbol)| {
+            state
+                .paths
+                .insert((location, imported), absolute_path.clone());
+            (state, absolute_path, symbol)
+        })
+}
+
+/// Checks an import
+///
+/// # Errors
+///
+/// This function will return an error if the imported path does not resolve to anything
+fn check_import<T: Clone>(
+    // the location at which the import is made
+    location: AbsoluteSymbolPath,
+    // the imported path, relative to the location
+    imported: Import,
+    // the current state
+    state: State<'_, T>,
+) -> Result<State<'_, T>, String> {
+    let (state, _, _) = check_path(location, imported.path, state)?;
+
+    Ok(state)
+}
+
+fn generate_path_map<T: FieldElement>(program: &ASMProgram<T>) -> Result<PathMap, String> {
+    check_module(
+        // the location of the main module
+        AbsoluteSymbolPath::default(),
+        &program.main,
+        // an empty state starting from this module
+        State {
+            root: &program.main,
+            paths: Default::default(),
+        },
+    )
+    .map(|state| state.paths)
+}
+
+/// Checks a module
+///
+/// # Errors
+///
+/// This function will return an error if a name is not unique, or if any path in this module does not resolve to anything
+fn check_module<'a, T: Clone>(
+    location: AbsoluteSymbolPath,
+    module: &ASMModule<T>,
+    state: State<'a, T>,
+) -> Result<State<'a, T>, String> {
+    let _ = module.symbol_definitions().try_fold(
+        BTreeMap::<String, Option<Symbol<T>>>::default(),
+        |mut acc, SymbolDefinition { name, .. }| match acc.insert(name.clone(), None) {
+            Some(_) => Err(format!("duplicate name {name}")),
+            None => Ok(acc),
+        },
+    )?;
+
+    module
+        .symbol_definitions()
+        // start with the initial state
+        .try_fold(state, |state, SymbolDefinition { name, symbol }| {
+            // update the state
+            match symbol {
+                Symbol::Machine(m) => check_machine(location.clone().join(name.clone()), m, state),
+                Symbol::Module(m) => {
+                    let m = match m {
+                        Module::External(_) => unreachable!(),
+                        Module::Local(m) => m,
+                    };
+                    check_module(location.clone().join(name.clone()), m, state)
+                }
+                Symbol::Import(s) => check_import(location.clone(), s.clone(), state),
+            }
+        })
+}
+
+/// Checks a machine, checking the paths it contains, in particular paths to the types of submachines
+///
+/// # Errors
+///
+/// This function will return an error if any of the paths does not resolve to anything
+fn check_machine<'a, T: Clone>(
+    location: AbsoluteSymbolPath,
+    m: &Machine<T>,
+    state: State<'a, T>,
+) -> Result<State<'a, T>, String> {
+    // we check the path in the context of the parent module
```

Comment:
Because the machine does not create a namespace?

Reply:
exactly